CRM-20264, Added Card type and Pan Truncation form fields on Payment …

eileenmcnaughton opened the pull request:
…form

I deliberately excluded card_type because I think we need to review the appropriateness of
the field first. My issue is that the wrangling wound up really wierd & it was apparent the issue
is that card_type is 'Visa' whereas card_type_id is 1 - we should either rename the field
or store the word.

eileenmcnaughton commented:
@monishdeb @JoeMurray @pradpnayak this adds pan_truncation to the back office payment forms - at least Contribution & AdditionalPayment. It would be easy to add card_type at the form level but I didn't because I think we need to either rename card_type to card_type_id if we are going to store an integer, or store the string 'Visa'

monishdeb commented:
Reviewing and making additional changes. Marking with WIP

eileenmcnaughton commented:
cool

eileenmcnaughton commented:
@monishdeb just letting you know - I just rebased this over master. I have committed to master a portion of @pradpnayak's pr that included the unit test for the additional payment form. I have added a commit to this PR that enables the rest of his unit test & makes it work. I used slightly different fixes for it though (called beginPostProcess & assignPaymentRelatedVariables) 

There is still further work in Pradeep's rewrite of the form that needs pulling in & review - but probably largely after this one. I was going to take a look at the tidy up of the send -email code next & try to separate that out & check how it was consolidating with functions on related forms

eileenmcnaughton commented:
@monishdeb note that I would suggesting merging this at the point when it is adding pan_truncation in a non-regressive way - ie. other forms are still working & Contribution & Add Payment are saving the new field. I think this is blocking the other clean  up on add payment form now

monishdeb commented:
@eileenmcnaughton I know, and I have already covered that in my local. Will soon push my commit with changes to support these two fields in all backoffice form, ensuring that it will not affect backoffice live mode and online registrations.

eileenmcnaughton commented:
Ok cool - sounds like you are progressing well!

eileenmcnaughton commented:
@monishdeb  this is too late for me to look at tonight- I'll try to tomorrow.

A couple of assumptions I had though
1) I was expecting us to just add card_type as a field & transform it after submit, to keep in line with front-end forms (preferably somewhere close to the form submit but acceptably deep in the BAO with some comments to suggest maybe moving it somewhere else later
2) I wasn't expecting to add those fields to the edit form for contributions. I think we should at some point make it possible to edit a transaction but given that it's possible to have multiple payments against a contribution they should not be editable as part of the contribution. I was expecting at some later point to figure out that editability side of it & also move the check_number over there.

Can we pull the edit part out for now & not try to get that merged into the rc because there is no way I can evaluate that part in time for the rc & I'm not sold it's the right answer for how to edit those fields

monishdeb commented:
@eileenmcnaughton I have added those two fields in backoffice form and here's the screenshot for each:
1. New Contribution backoffice form
![image](https://cloud.githubusercontent.com/assets/3735621/25180480/c7264066-252b-11e7-81a1-089fe62fda93.png)

2. Edit Contribution backoffice form (although I have freezed those two element if payment was done on using any ```payment_processor_id !=0```, but still the field should be readonly instead of hidden)
![image](https://cloud.githubusercontent.com/assets/3735621/25180576/29212d12-252c-11e7-8ed3-f50d4bbdfb0f.png)

3. Membership backoffice form
![image](https://cloud.githubusercontent.com/assets/3735621/25180599/423ecc6e-252c-11e7-86f9-745d2c7f3844.png)

4. Event registration backoffice form
![image](https://cloud.githubusercontent.com/assets/3735621/25180637/646cfd9c-252c-11e7-9a48-3b4f9dac3031.png)

monishdeb commented:
@eileenmcnaughton issues which I encountered till now are as follows:
1. On Membership and Event backoffice form submission, I don't get any values for those two fields. I think the issue is related to https://github.com/civicrm/civicrm-core/pull/10126 
2. The frozen fields on Edit contribution form should be readonly not hidden. But this [code](https://github.com/civicrm/civicrm-core/pull/10157/files#diff-e4bae7918cbe2efb1ea9e98c7503251aR346) is not working due to unexpected reason. 

Regarding your queries, please find my inline explanation:
> I was expecting us to just add card_type as a field & transform it after submit, to keep in line with front-end forms (preferably somewhere close to the form submit but acceptably deep in the BAO with some comments to suggest maybe moving it somewhere else later
I tried to use the similar [credit_card_type](https://github.com/civicrm/civicrm-core/pull/10157/files#diff-f8705625754a6a2f3655a8c7c696969cR693) field on backoffice offline form but I encountered some JS issues as we also need to call ```CRM_Financial_Form_Payment::addCreditCardJs(...)``` on CRM_Contribution_Form_Conribution::buildQuickForm() to display credit card type icon like in live mode/online registration. Wil give it another try. 

> I wasn't expecting to add those fields to the edit form for contributions. I think we should at some point make it possible to edit a transaction but given that it's possible to have multiple payments against a contribution they should not be editable as part of the contribution. I was expecting at some later point to figure out that editability side of it & also move the check_number over there.
Actually that won't be under the scope of IIDA-114 task, so we need to create a backlog ticket in JIRA to handle it in future.

monishdeb commented:
This will be replacement for #10050, #10051 and #10052

eileenmcnaughton commented:
Those reasons are kind of why I wanted to get a PR merged that only worried about the pan_truncation & avoided the edit form - that seems like a discreet change that gets us a piece of functionality.  I think there is a big issue across all payment forms that they should be, and are not, calling a common function in  a common way for processing payment parameters - this has also led to multiple attempts to ensure the email is passed to the payment processor.

Can we leave off including those fields at all on the edit form for now? I don't think they should be present but frozen - I think we should find another way to present details for a payment within a contribution for editing & leave that until later

eileenmcnaughton commented:
Re card type id - the thing is on back-office forms we will already be presenting card_type for credit card present payments. So if we fix to ensure that it is always saved for when we do back-office credit card payments then it will 'just save' for the manual payment processor (pay later) ones. Hence I was thinking we should ensure that was being saved before worrying about the manual entry

eileenmcnaughton commented on CRM/Contribute/Form/Contribution.php:
```diff
@@ -1268,13 +1268,22 @@ protected function processCreditCard($submittedValues, $lineItem, $contactID) {
+              'pan_truncation' => CRM_Utils_Array::value('pan_truncation', $paymentParams),
+            );
+            if (!empty($paymentParams['credit_card_type'])) {
+              $params['card_type_id'] = CRM_Core_BAO_FinancialTrxn::formatCreditCardDetails($paymentParams['credit_card_type']);
```
so we shouldn't need to do handling here if we have called beginPostProcess

eileenmcnaughton commented on CRM/Contribute/Form/Contribution/Confirm.php:
```diff
@@ -2440,6 +2440,8 @@ protected function completeTransaction($result, $contributionID) {
             'is_transactional' => FALSE,
             'fee_amount' => CRM_Utils_Array::value('fee_amount', $result),
             'receive_date' => CRM_Utils_Array::value('receive_date', $result),
+            'card_type_id' => CRM_Core_BAO_FinancialTrxn::formatCreditCardDetails(CRM_Utils_Array::value('credit_card_type', $result)),
```
we shouldn't need this if we call beginPostProcess (or similar as I think this form doesn't share that )

eileenmcnaughton replied:
let's get this happening early on

eileenmcnaughton commented on CRM/Core/Payment/Form.php:
```diff
@@ -326,6 +326,27 @@ public static function setDefaultValues(&$form, $contactID) {
+    // In backoffice edit form, retrieve pan_truncation and card type of corresponding entity from latest payment,
+    //  if found that set in default values. However if the payment was done using any payment processor either via
+    //  backoffice credit card or online registration, freeze these two fields so user cannot edit their corresponding values
+    if (!empty($form->isBackOffice) && !empty($form->_entityID)) {
```
so I think all this should be excluded at this stage

monishdeb replied:
At least this won't cause any regression and is required by IIDA client. So what if we allow it in 4.7.19 and handle it in next phase may be 4.7.20. We could leave a todo note here as reminder. What ya say ?

eileenmcnaughton commented on CRM/Core/Payment/Form.php:
```diff
@@ -424,4 +445,30 @@ public static function getCreditCardExpirationYear($src) {
+   *
+   * @return array
+   */
+  public static function getlatestPayment($entityID, $entity) {
```
exclude this until we have a plan for a transaction edit form

eileenmcnaughton commented on CRM/Event/Form/Participant.php:
```diff
@@ -212,6 +212,7 @@ public function preProcess() {
     $this->_showFeeBlock = CRM_Utils_Array::value('eventId', $_GET);
     $this->assign('showFeeBlock', FALSE);
     $this->assign('feeBlockPaid', FALSE);
+    $this->assign('isBackOffice', $this->isBackOffice);
```
actually we don't need this - if we are called parent::preProcess

eileenmcnaughton commented on CRM/Financial/Form/Payment.php:
```diff
@@ -37,6 +37,7 @@ class CRM_Financial_Form_Payment extends CRM_Core_Form {
    */
   protected $_paymentProcessorID;
   protected $currency;
```
we don't need the changes in this form if we are not trying to overload the payment form with the edit transaction form

eileenmcnaughton commented on api/v3/Contribution.php:
```diff
@@ -160,6 +160,14 @@ function _civicrm_api3_contribution_create_spec(&$params) {
     'type' => CRM_Utils_Type::T_STRING,
     'description' => 'Transaction ID specific to the refund taking place',
   );
+  $params['card_type_id'] = array(
```
yes

eileenmcnaughton commented on templates/CRM/Contribute/Form/Contribution.tpl:
```diff
@@ -235,6 +235,10 @@
   </table>
 
   {if !$contributionMode}
+    <fieldset class="payment-details_group">
```
I actually killed this because I thought it was ugly...

eileenmcnaughton commented on templates/CRM/common/paymentBlock.tpl:
```diff
@@ -104,12 +104,19 @@
       {else}
         {capture assign='profilePathVar'}{/capture}
       {/if}
+      {if $contribID}
```
I don't think we need any of the changes on this form if we are not doing the edit overload thing

eileenmcnaughton commented on tests/phpunit/CRM/Core/BAO/FinancialTrxnTest.php:
```diff
@@ -244,4 +244,14 @@ public function testUpdateCreditCardDetails() {
+  /**
+   * Test CRM_Core_BAO_FinancialTrxn::formatCreditCardDetails(..)
+   */
+  public function testFormatCreditCardDetails() {
```
I'm kind of hopeful the formatCreditCardDetails function might be dead by the time this is merged

eileenmcnaughton commented on tests/phpunit/CRM/Contribute/Form/ContributionTest.php:
```diff
@@ -922,4 +922,101 @@ public function testReSubmitSaleTax() {
+  /**
+   * function to test card_type and pan truncation.
+   */
+  public function testCardTypeAndPanTruncation() {
```
this is good.

monishdeb commented on CRM/Contribute/Form/Contribution/Confirm.php:
```diff
@@ -2037,6 +2037,8 @@ protected function processFormSubmission($contactID) {
     // fix currency ID
     $this->_params['currencyID'] = CRM_Core_Config::singleton()->defaultCurrency;
 
+    CRM_Contribute_Form_AbstractEditPayment::formatCreditCardDetails($this->_params);
```
@eileenmcnaughton we cannot call beginPostProcess() here as it extend different parent class also there are lots of formvalue attributes which are formatted here (and to move into common formatting function will need lots of refactoring changes) so I thought its better format the credit card details only with the help of this function.

eileenmcnaughton replied:
OK cool

monishdeb commented:
@eileenmcnaughton I have made the final changes https://github.com/civicrm/civicrm-core/pull/10157/commits/3cd1f88a2b540c37b0ab19c9ac28c94358a4f550 Have a look

eileenmcnaughton commented on CRM/Core/Payment.php:
```diff
@@ -61,6 +61,18 @@
+   * @todo create option group - nb omnipay uses a 3rd type - transparent redirect cc
+   */
+  const
+    PAYMENT_TYPE_CREDIT_CARD = 1,
```
I think this got re-added by accident?

eileenmcnaughton replied:
Ah I see - still required from an ugly flow

eileenmcnaughton replied:
ok

eileenmcnaughton commented on templates/CRM/common/paymentBlock.tpl:
```diff
@@ -133,6 +133,13 @@
     });
 
+    // no need to show `pan_truncation` and `credit_card_type` fields on live mode
+    {/literal}{if $isBackOffice}{literal};
```
Why do we need this? the fact isBackOffice gets passed is not enough?

monishdeb replied:
Yes its the criteria that differentiate backoffice live and offline mode

eileenmcnaughton commented:
Test error is PHP Fatal error:  Call to undefined method CRM_Core_BAO_FinancialTrxn::formatCreditCardDetails() in /home/jenkins/buildkit/build/core-10157-92n7k/sites/all/modules/civicrm/tests/phpunit/CRM/Contribute/Form/ContributionTest.php on line 967

eileenmcnaughton commented:
As this PR stands right now there is a minor test issue (just an update to the test class) and I'm not sure about this line

https://github.com/civicrm/civicrm-core/pull/10157/files#diff-726c3dcc8c2364a0b9b8f0c0e11bcd34


Apart from that there is more testing to do on the membership form etc - but I'm OK to keep testing those things after merge during the rc period

eileenmcnaughton commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -3135,6 +3135,8 @@ public static function recordFinancialAccounts(&$params, $financialTrxnValues =
         $balanceTrxnParams['payment_instrument_id'] = $params['contribution']->payment_instrument_id;
         $balanceTrxnParams['check_number'] = CRM_Utils_Array::value('check_number', $params);
+        $balanceTrxnParams['pan_truncation'] = CRM_Utils_Array::value('pan_truncation', $params);
+        $balanceTrxnParams['card_type_id'] = CRM_Utils_Array::value('card_type_id', $params);
```
ok

eileenmcnaughton commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -3199,6 +3201,8 @@ public static function recordFinancialAccounts(&$params, $financialTrxnValues =
         'payment_instrument_id' => CRM_Utils_Array::value('payment_instrument_id', $params, $params['contribution']->payment_instrument_id),
         'check_number' => CRM_Utils_Array::value('check_number', $params),
+        'pan_truncation' => CRM_Utils_Array::value('pan_truncation', $params),
+        'card_type_id' => CRM_Utils_Array::value('card_type_id', $params),
```
ok

eileenmcnaughton commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -4464,6 +4469,8 @@ public static function completeOrder(&$input, &$ids, $objects, $transaction, $re
       'receipt_date',
       'contribution_status_id',
+      'card_type_id',
+      'pan_truncation',
```
ok

eileenmcnaughton commented on CRM/Contribute/Form/AbstractEditPayment.php:
```diff
@@ -221,6 +221,7 @@ public function preProcess() {
     CRM_Core_Resources::singleton()->addVars('coreForm', array('contact_id' => (int) $this->_contactID));
     $this->_action = CRM_Utils_Request::retrieve('action', 'String', $this, FALSE, 'add');
     $this->_mode = empty($this->_mode) ? CRM_Utils_Request::retrieve('mode', 'String', $this) : $this->_mode;
+    $this->assign('isBackOffice', $this->isBackOffice);
```
ok

eileenmcnaughton commented on CRM/Contribute/Form/AbstractEditPayment.php:
```diff
@@ -573,15 +574,27 @@ protected function beginPostProcess() {
-      $this->_params['pan_truncation'] = substr($this->_params['credit_card_number'], -4);
-    }
 
+    self::formatCreditCardDetails($this->_params);
```
ok

eileenmcnaughton commented on CRM/Contribute/Form/AbstractEditPayment.php:
```diff
@@ -573,15 +574,27 @@ protected function beginPostProcess() {
+   *
+   * @return void
+   */
+  public static function formatCreditCardDetails(&$params) {
```
ok

eileenmcnaughton commented on CRM/Contribute/Form/Contribution.php:
```diff
@@ -1274,6 +1274,8 @@ protected function processCreditCard($submittedValues, $lineItem, $contactID) {
               'payment_processor_id' => $this->_paymentProcessor['id'],
               'is_transactional' => FALSE,
               'fee_amount' => CRM_Utils_Array::value('fee_amount', $result),
+              'card_type_id' => CRM_Utils_Array::value('card_type_id', $paymentParams),
```
ok

eileenmcnaughton commented on CRM/Contribute/Form/Contribution.php:
```diff
@@ -1643,6 +1646,8 @@ protected function submit($submittedValues, $action, $pledgePaymentID) {
         'source',
         'check_number',
+        'card_type_id',
+        'pan_truncation',
```
ok

eileenmcnaughton commented on CRM/Contribute/Form/Contribution/Confirm.php:
```diff
@@ -2434,14 +2436,15 @@ protected function completeTransaction($result, $contributionID) {
+          'receive_date' => CRM_Utils_Array::value('receive_date', $result),
+          'card_type_id' => CRM_Utils_Array::value('card_type_id', $result),
+          'pan_truncation' => CRM_Utils_Array::value('pan_truncation', $result),
+        ));
```
ok

eileenmcnaughton commented on CRM/Core/Payment.php:
```diff
@@ -781,6 +793,13 @@ public function getPaymentFormFieldsMetadata() {
+            'rule_name' => 'numeric',
+            'rule_parameters' => NULL,
+          ),
+        ),
```
ok

eileenmcnaughton commented on CRM/Core/Payment/Form.php:
```diff
@@ -67,8 +67,8 @@ static public function setPaymentFieldsByProcessor(&$form, $processor, $billing_
     $paymentTypeName = self::getPaymentTypeName($processor);
-    $paymentTypeLabel = self::getPaymentTypeLabel($processor);
     $form->assign('paymentTypeName', $paymentTypeName);
+    $paymentTypeLabel = self::getPaymentTypeLabel($processor);
```
ok

eileenmcnaughton commented on CRM/Core/Payment/Manual.php:
```diff
@@ -95,7 +95,7 @@ public function getPaymentFormFields() {
       // credit_card_type_id, or it should store 'visa' It probably makes sense to fix that before going much
       // further as the code I've seen makes it clear that it will require work arounds.
-      return array('pan_truncation');
+      return array('credit_card_type', 'pan_truncation');
```
ok

eileenmcnaughton commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2408,6 +2408,8 @@ public static function recordMembershipContribution(&$params, $ids = array()) {
       'skipLineItem',
       'contribution_recur_id',
+      'pan_truncation',
+      'card_type_id',
```
ok

eileenmcnaughton commented on CRM/Member/Form/Membership.php:
```diff
@@ -1308,6 +1308,8 @@ public function submit() {
         'campaign_id',
         'receive_date',
+        'card_type_id',
+        'pan_truncation',
```
ok

eileenmcnaughton commented:
@monishdeb  I have squashed all commits into a single one and removed that one line that I have not yet understood & setting to merge-on-pass. Once I understand the reason for that line we can fix - in the rc if need be

seamuslee001 commented:
pinging @eileenmcnaughton @monishdeb this has passed testing now

monishdeb commented:
Cool

monishdeb commented:
@eileenmcnaughton @JoeMurray the only left in this issue is showing the default values in edit form. Should I create a backlog ticket in JIRA for that ?

eileenmcnaughton commented:
My thinking here is that is makes sense to have a transaction-edit form. Since it is a new form I would probably do it in angular - which is possibly a learning curve, but an important one, and make it accessible from a transaction level action menu. 

![screenshot 2017-04-23 11 33 47](https://cloud.githubusercontent.com/assets/336308/25309129/c0159c32-2818-11e7-886d-92eb63e9601f.png)

We could check with @colemanw whether he thinks something inline edity would make sense.  I would think that to do something like this in an extension in a non-override way - with a view to moving to core would probably be fairly quick for someone like Coleman who is used to angular but a little slower for someone else.

Eventually we'd probably move check_number  & trxn_id out of the contribution table altogether. I would also say that while I'm not opposed to locking the card_type when processor_id exists I would probably not lock it - on the basis that actually a lot of those transactions have been imported or entered the system in a variety of ways so the argument 'it has been set by the processor so don't let users edit it' is not always true

JoeMurray commented:
Sounds like a nice improvement. Could you create an issue for the new transaction edit form? What is the workflow that would require editing the system level table? Should we perhaps consider this as part of editting a payment, which is a major use of this table? Brian was thinking we might benefit from splitting financial_transaction into separate tables for payments and other accounting entries.

While we considered removing civicrm_contribution.trxn_id in 4.3 we were worried about users having found a use for it out in the wild. contribution.check_number definitely should not exist on contribution form or table any more...more cruft which exists probably because of worry about upgrade issues changing where the data appears in the interface.

JoeMurray commented:
@colemanw and @eileenmcnaughton for now I think we are ready to merge and say we're done with this round of improvements.

JoeMurray commented:
Oops. I see this has been merged!

eileenmcnaughton commented:
@JoeMurray yep this is merged.

The idea about the transaction edit page was to allow people to edit pan_truncation & card_type (& also check_number) after the initial submit.
